File: scripts/verify-paid-exit-seller-ui-receipts.py
#!/usr/bin/env python3
"""Validate shipped-UI paid-exit seller receipts on supported platforms."""

from __future__ import annotations

import argparse
import hashlib
import json
import pathlib
import re


EXPECTED_PLATFORMS = {"linux", "macos"}


def digest(path: pathlib.Path) -> str:
    return hashlib.sha256(path.read_bytes()).hexdigest()


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--app-git-sha", required=True)
    parser.add_argument("--app-git-tree", required=True)
    parser.add_argument("--output", required=True)
    parser.add_argument("receipts", nargs="+")
    args = parser.parse_args()
    if not re.fullmatch(r"[0-9a-f]{40}", args.app_git_sha):
        parser.error("--app-git-sha must be an exact commit")
    if not re.fullmatch(r"[0-9a-f]{40}", args.app_git_tree):
        parser.error("--app-git-tree must be an exact tree")

    paths: dict[str, pathlib.Path] = {}
    for raw in args.receipts:
        platform, separator, name = raw.partition("=")
        if not separator or platform not in EXPECTED_PLATFORMS:
            parser.error(f"invalid platform=receipt: {raw}")
        if platform in paths:
            parser.error(f"duplicate receipt for {platform}")
        paths[platform] = pathlib.Path(name)
    if set(paths) != EXPECTED_PLATFORMS:
        parser.error(
            f"expected receipts for {sorted(EXPECTED_PLATFORMS)}, got {sorted(paths)}"
        )

    hashes: dict[str, str] = {}
    for platform, path in sorted(paths.items()):
        value = json.loads(path.read_text(encoding="utf-8"))
        expected = {
            "receiptSchema": 1,
            "platform": platform,
            "case": "paid-exit-seller",
            "evidenceSource": "shipped-ui-restart-readback",
            "releaseBlackbox": True,
            "savedViaShippedUi": True,
            "enabledViaShippedUi": True,
            "uiRestartReadback": True,
            "privateStateRead": False,
            "paidExitEnabled": True,
            "paidExitPriceMsatPerGb": 1_000_000,
            "paidExitCountryCode": "FI",
            "paidExitAcceptedMints": ["http://cashu-mint:3338"],
            "appGitSha": args.app_git_sha,
            "appGitTree": args.app_git_tree,
        }
        for field, wanted in expected.items():
            if value.get(field) != wanted:
                raise ValueError(
                    f"{platform} receipt {field}: expected {wanted!r}, "
                    f"got {value.get(field)!r}"
                )
        hashes[platform] = digest(path)

    output = pathlib.Path(args.output)
    output.parent.mkdir(parents=True, exist_ok=True)
    temporary = output.with_name(f".{output.name}.tmp")
    temporary.write_text(
        json.dumps(
            {
                "receiptSchema": 1,
                "gate": "supported-platform paid-exit seller shipped UI",
                "platforms": hashes,
                "allSupportedPlatformsSavedEnabledAndRestartRead": True,
                "sellerContract": {
                    "priceMsatPerGb": 1_000_000,
                    "countryCode": "FI",
                    "acceptedMints": ["http://cashu-mint:3338"],
                },
                "appGitSha": args.app_git_sha,
                "appGitTree": args.app_git_tree,
            },
            indent=2,
            sort_keys=True,
        )
        + "\n",
        encoding="utf-8",
    )
    temporary.replace(output)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/verify-paid-exit-seller-ui-receipts.mjs
#!/usr/bin/env node
// Validate shipped-UI paid-exit seller receipts using the canonical source proof.
import { createHash } from 'node:crypto'
import { lstatSync, mkdirSync, readFileSync, renameSync, writeFileSync } from 'node:fs'
import { dirname, resolve } from 'node:path'
import { fileURLToPath } from 'node:url'
import { isDeepStrictEqual, parseArgs } from 'node:util'
import { proveUnchangedPlatformInputs } from './release-component-source.mjs'

const EXPECTED_PLATFORMS = ['linux', 'macos']
const { values, positionals } = parseArgs({
  allowPositionals: true,
  options: Object.fromEntries(
    ['candidate-root', 'app-git-sha', 'app-git-tree', 'output'].map((name) => [name, { type: 'string' }]),
  ),
})
const candidateRoot = resolve(values['candidate-root'] ?? fileURLToPath(new URL('..', import.meta.url)))
const candidateCommit = values['app-git-sha'] ?? ''
const candidateTree = values['app-git-tree'] ?? ''
if (!/^[0-9a-f]{40}$/.test(candidateCommit) || !/^[0-9a-f]{40}$/.test(candidateTree) || !values.output) {
  throw new Error('Exact --app-git-sha, --app-git-tree and --output are required.')
}
const paths = new Map()
for (const raw of positionals) {
  const separator = raw.indexOf('=')
  const platform = raw.slice(0, separator)
  if (separator < 0 || !EXPECTED_PLATFORMS.includes(platform) || paths.has(platform)) {
    throw new Error(`Invalid or duplicate platform=receipt: ${raw}`)
  }
  paths.set(platform, raw.slice(separator + 1))
}
if (!isDeepStrictEqual([...paths.keys()].sort(), EXPECTED_PLATFORMS)) {
  throw new Error('Expected receipts for linux and macos.')
}

const sellerContract = {
  priceMsatPerGb: 1_000_000,
  countryCode: 'FI',
  acceptedMints: ['http://cashu-mint:3338'],
}
const platforms = {}
const platformSourceEquivalence = {}
for (const platform of EXPECTED_PLATFORMS) {
  const path = paths.get(platform)
  if (!lstatSync(path).isFile()) throw new Error(`${platform} receipt must be a regular file.`)
  const bytes = readFileSync(path)
  const value = JSON.parse(bytes)
  const expected = {
    receiptSchema: 1, platform, case: 'paid-exit-seller',
    evidenceSource: 'shipped-ui-restart-readback', releaseBlackbox: true,
    savedViaShippedUi: true, enabledViaShippedUi: true, uiRestartReadback: true,
    privateStateRead: false, paidExitEnabled: true,
    paidExitPriceMsatPerGb: sellerContract.priceMsatPerGb,
    paidExitCountryCode: sellerContract.countryCode,
    paidExitAcceptedMints: sellerContract.acceptedMints,
  }
  for (const [field, wanted] of Object.entries(expected)) {
    if (!isDeepStrictEqual(value[field], wanted)) throw new Error(`${platform} receipt ${field} differs from the seller UI contract.`)
  }
  // Always resolve both trees, including exact-source receipts. Never rewrite
  // old evidence with the new commit: retain its digest and record the proof.
  const proof = proveUnchangedPlatformInputs({
    candidateRoot, platform, candidateCommit, candidateTree,
    receiptCommit: value.appGitSha, receiptTree: value.appGitTree,
  })
  if (value.appGitSha !== candidateCommit) platformSourceEquivalence[platform] = proof
  platforms[platform] = createHash('sha256').update(bytes).digest('hex')
}
const output = resolve(values.output)
mkdirSync(dirname(output), { recursive: true })
const temporary = `${output}.tmp.${process.pid}`
writeFileSync(temporary, `${JSON.stringify({
  receiptSchema: 1,
  gate: 'supported-platform paid-exit seller shipped UI',
  platforms, platformSourceEquivalence,
  allSupportedPlatformsSavedEnabledAndRestartRead: true,
  sellerContract, appGitSha: candidateCommit, appGitTree: candidateTree,
}, null, 2)}\n`, { flag: 'wx' })
renameSync(temporary, output)
